tests/protocol/test_wire.py

```diff
@@ -516,227 +521,235 @@ def test_get_ext_conf_with_uri(self, *args):
+            original_http_request = restutil.http_request
+
+            def http_request(method, url, *args, **kwargs):
+                if method == 'GET':
+                    if url == extension_url:
+                        http_request.urls.append(url)
+                        raise HttpError("Exception to fake an error on the direct channel")
+                    elif TestWireClient._is_extension_artifact_host_request(extension_url, url, **kwargs):
+                        http_request.urls.append(url)
+                        return MockResponse(body=b'', status_code=200)
+                return original_http_request(method, url, *args, **kwargs)
+            http_request.urls = []
+
+        with patch("azurelinuxagent.common.utils.restutil.http_request", side_effect=http_request):
+            HostPluginProtocol.set_default_channel(False)
+
+            success = protocol.download_ext_handler_pkg(extension_url, target_file)
+
+            self.assertEquals(success, True, 'The download should have succeeded')
+            self.assertEquals(len(http_request.urls), 2, "Unexpected number of HTTP requests: [{0}]".format(http_request.urls))
+            self.assertEquals(http_request.urls[0], extension_url, "The first attempt should have been over the direct channel")
+            self.assertTrue(http_request.urls[1].endswith('/extensionArtifact'), "The retry attempt should have been over the host channel")
+            self.assertTrue(os.path.exists(target_file), 'The extension package was not downloaded')
+            self.assertEquals(HostPluginProtocol.is_default_channel(), True, "The host channel should have been set as the default")
+
+    def test_download_ext_handler_pkg_should_retry_the_host_channel_after_refreshing_host_plugin(self):
+        extension_url = 'https://fake_host/fake_extension.zip'
+        target_file = os.path.join(self.tmp_dir, 'fake_extension.zip')
 
-        # Direct channel fails, host channel succeeds. Goal state should not have been updated and host channel
-        # should have been set as default
-        with patch("azurelinuxagent.common.utils.restutil._http_request",
-                   side_effect=[mock_failed_response, mock_successful_response]):
-            with patch("azurelinuxagent.common.protocol.wire.WireClient.update_goal_state") as mock_update_goal_state:
-                with patch("azurelinuxagent.common.protocol.wire.WireClient.fetch", wraps=client.fetch) as patch_direct:
-                    with patch("azurelinuxagent.common.protocol.wire.WireClient.fetch_manifest_through_host",
-                               wraps=client.fetch_manifest_through_host) as patch_host:
-                        ret = client.fetch_manifest([VMAgentManifestUri(uri="uri1")])
-                        self.assertEquals(ret, "OK")
+        with mock_wire_protocol.create(mockwiredata.DATA_FILE) as protocol:
+            protocol.client.get_host_plugin()  # force initialization of the host plugin
```

Comment:
Why is this needed here and not in the test above?

Reply:
It is not needed, but simplifies the mock below. Initialization of the host plugin does a request for the goal state. Since the mock below is tracking goal state requests I would need to add code to filter out this one request. It is not too hard, but adding that code there would make the code less readable since it would distract from the actual intention of the test.

Some of the other tests don't track goal state requests so I do it only when it simplifies the code.

I'll add a coment.